src/h5web/visualizations/shared/utils.ts

```diff
@@ -45,33 +45,37 @@ export function computeVisSize(
     : { width, height: width / aspectRatio };
 }
 
-export function findDomain(data: number[]): Domain | undefined {
-  const domain = extent(data);
-  return domain[0] !== undefined && domain[1] !== undefined
-    ? domain
-    : undefined;
-}
-
-export function getSupportedDomain(
-  domain: Domain | undefined,
-  scaleType: ScaleType,
-  values: number[]
+export function getDataDomain(
+  values: number[],
+  scaleType: ScaleType = ScaleType.Linear
 ): Domain | undefined {
-  if (!domain) {
+  if (values.length === 0) {
     return undefined;
   }
 
-  // If scale is not log or domain does not cross zero, domain is supported
-  if (scaleType !== ScaleType.Log || domain[0] * domain[1] > 0) {
-    return domain;
+  const [min, max, positiveMin] = values.reduce<
+    [number, number, number | undefined]
+  >(
+    (acc, val) => {
+      const [accMin, accMax, accPositiveMin] = acc;
+      return [
+        accMin > val ? val : accMin,
+        accMax < val ? val : accMax,
+        val > 0 && (accPositiveMin === undefined || accPositiveMin > val)
+          ? val
+          : accPositiveMin,
+      ];
+    },
+    [values[0], values[0], undefined]
+  );
```

Comment:
This algorithm requires only one loop over the array, even when the domain has to be clamped. Before, we used to call `findDomain`, and therefore `d3.extent`, twice.

Reply:
Nice !